src/memory.rs

```diff
@@ -153,17 +167,21 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
             return Err(EvalError::Unimplemented(format!("bad pointer offset: {}", ptr.offset)));
         }
         if ptr.points_to_zst() {
-            return Ok(self.allocate(new_size));
+            return self.allocate(new_size);
         }
 
         let size = self.get_mut(ptr.alloc_id)?.bytes.len();
 
         if new_size > size {
             let amount = new_size - size;
+            self.memory_usage += amount as u64;
             let alloc = self.get_mut(ptr.alloc_id)?;
             alloc.bytes.extend(iter::repeat(0).take(amount));
             alloc.undef_mask.grow(amount, false);
         } else if size > new_size {
+            // it's possible to cause miri to use arbitrary amounts of memory that aren't detectable
```

Comment:
Actually, come to think of it, it might be sensible to do that unconditionally whenever a downsizing happens. The only code that does this is stuff like `Vec` which rarely gets downsized in practice.